File: uspec/templates/frame.js
// Build one uSpec template as a detached FRAME, next to the component being documented.
//
// This replaces Step "Import and Detach Template" in every `create-*` skill. That step reads:
//
//     const templateComponent = await figma.importComponentByKeyAsync(TEMPLATE_KEY);
//     const instance = templateComponent.createInstance();
//     const frame = instance.detachInstance();
//
// — it imports a component, instantiates it, and throws the component away on the next line.
// Everything downstream works on the plain frame and finds its parts by layer name. So the
// template never needs to exist as a component in the file at all: building the same node tree
// from the captured description produces the identical frame, with the identical `#anchor`
// layers, and skips the import that hangs for an unpublished cross-file key.
//
// Returns the same shape the original step returns: { frameId, pageId, pageName }.
//
// Constants the caller injects:
//   SPEC              one entry of templates.json -> templates (e.g. the "API" object)
//   COMPONENT_NODE_ID the component being documented; the frame is placed to its right
//   FRAME_NAME        e.g. "my-button API"
//   GAP               optional, defaults to 200

const _GAP = typeof GAP === "number" ? GAP : 200;

const compNode = await figma.getNodeByIdAsync(COMPONENT_NODE_ID);
if (!compNode) throw new Error("component " + COMPONENT_NODE_ID + " not found in " + figma.root.name);

// The page the component lives on. Figmosha inherits Figma Desktop's page context and does not
// need the setCurrentPageAsync walk-up the MCP columns use, but the frame still has to be
// appended to the component's own page rather than to whatever page happens to be current.
let _page = compNode;
while (_page.parent && _page.parent.type !== "DOCUMENT") _page = _page.parent;
if (_page.type !== "PAGE") throw new Error("could not find the page holding " + COMPONENT_NODE_ID);

// ---------------------------------------------------------------- fonts

const _fontKeys = {};
const _collectFonts = (n) => {
  if (n.type === "TEXT") {
    if (n.fontName && n.fontName !== "__MIXED__") _fontKeys[n.fontName.family + "|" + n.fontName.style] = 1;
    for (const r of n.runs || []) _fontKeys[r.family + "|" + r.style] = 1;
  }
  for (const c of n.children || []) _collectFonts(c);
};
_collectFonts(SPEC);
for (const k of Object.keys(_fontKeys)) {
  const f = { family: k.split("|")[0], style: k.split("|")[1] };
  try {
    await figma.loadFontAsync(f);
  } catch (e) {
    throw new Error("font unavailable in " + figma.root.name + ": " + f.family + " " + f.style);
  }
}

// ---------------------------------------------------------------- build
//
// Four passes, because Figma will not take them in one: a child's layoutSizing is only
// settable once its parent already has a layoutMode.

const _SKIP = {
  type: 1, name: 1, children: 1, width: 1, height: 1, x: 1, y: 1,
  characters: 1, fontName: 1, runs: 1, vectorPaths: 1, fills: 1, strokes: 1, effects: 1,
  layoutMode: 1, layoutWrap: 1, itemSpacing: 1, counterAxisSpacing: 1, itemReverseZIndex: 1,
  paddingTop: 1, paddingRight: 1, paddingBottom: 1, paddingLeft: 1,
  primaryAxisSizingMode: 1, counterAxisSizingMode: 1,
  primaryAxisAlignItems: 1, counterAxisAlignItems: 1,
  layoutSizingHorizontal: 1, layoutSizingVertical: 1,
  layoutAlign: 1, layoutGrow: 1, layoutPositioning: 1,
  strokeWeight: 1, overflowDirection: 1, autoRename: 1,
};

// Layout properties are load-bearing: a swallowed failure here is a frame that looks right
// and lays out wrong. They are reported; everything else may fail quietly, because the
// capture carries properties that simply do not exist on every node type.
const _failures = [];
const _LOUD = { layoutSizingHorizontal: 1, layoutSizingVertical: 1, layoutAlign: 1,
                layoutGrow: 1, layoutMode: 1, layoutPositioning: 1 };

const _set = (node, prop, value) => {
  if (value === "__MIXED__" || value === undefined || value === null) return;
  try {
    node[prop] = value;
  } catch (e) {
    if (_LOUD[prop])
      _failures.push({ node: node.name, type: node.type, property: prop, value: value,
                       message: String(e && e.message) });
  }
};

function _build(spec, isRoot) {
  // The root is a FRAME, not a COMPONENT: the original step detaches to a frame anyway.
  let n;
  if (isRoot || spec.type === "FRAME" || spec.type === "COMPONENT") n = figma.createFrame();
  else if (spec.type === "TEXT") n = figma.createText();
  else if (spec.type === "VECTOR") n = figma.createVector();
  else throw new Error("unexpected node type in the captured template: " + spec.type);

  n.name = spec.name;

  if (spec.type === "TEXT") {
    if (spec.fontName && spec.fontName !== "__MIXED__") n.fontName = spec.fontName;
    n.characters = spec.characters || "";
    if (spec.runs) for (const r of spec.runs)
      n.setRangeFontName(r.start, r.end, { family: r.family, style: r.style });
  }
  if (spec.type === "VECTOR" && spec.vectorPaths) n.vectorPaths = spec.vectorPaths;

  if (Array.isArray(spec.fills)) _set(n, "fills", spec.fills);
  if (Array.isArray(spec.strokes)) _set(n, "strokes", spec.strokes);
  if (Array.isArray(spec.effects)) _set(n, "effects", spec.effects);

  for (const [k, v] of Object.entries(spec)) if (!_SKIP[k]) _set(n, k, v);
  for (const c of spec.children || []) n.appendChild(_build(c, false));
  return n;
}

function _layout(node, spec) {
  if (spec.layoutMode && spec.layoutMode !== "NONE") {
    node.layoutMode = spec.layoutMode;
    for (const p of ["layoutWrap", "itemSpacing", "counterAxisSpacing", "primaryAxisAlignItems",
                     "counterAxisAlignItems", "itemReverseZIndex"]) _set(node, p, spec[p]);
  }
  // Padding regardless of layoutMode. Two Motion frames (#ruler-track, #track-area) keep
  // padding 12/8/12/8 with layoutMode NONE — Figma retains the values after auto-layout is
  // switched off. Inert visually, but carrying them makes the capture round-trip exactly.
  for (const p of ["paddingTop", "paddingRight", "paddingBottom", "paddingLeft"])
    _set(node, p, spec[p]);
  const kids = node.children || [];
  (spec.children || []).forEach((cs, i) => { if (kids[i]) _layout(kids[i], cs); });
}

function _sizing(node, spec) {
  _set(node, "layoutPositioning", spec.layoutPositioning);

  // Resize FIRST. resize() writes both axes, which silently converts a HUG axis to FIXED — so
  // giving the node its captured size has to happen before the sizing modes are declared, not
  // after. This is what made the root frame come back FIXED on the first build.
  if (spec.layoutSizingHorizontal === "FIXED" || spec.layoutSizingVertical === "FIXED") {
    try { node.resize(spec.width, spec.height); } catch (e) {}
  }

  // FILL and STRETCH are the same statement in two vocabularies, and Figma rejects the newer
  // one on a node whose parent axis does not support it. Set layoutAlign first so STRETCH
  // carries the intent, then let layoutSizing* refine it where it is accepted.
  _set(node, "layoutAlign", spec.layoutAlign);
  _set(node, "layoutGrow", spec.layoutGrow);
  _set(node, "layoutSizingHorizontal", spec.layoutSizingHorizontal);
  _set(node, "layoutSizingVertical", spec.layoutSizingVertical);

  const kids = node.children || [];
  (spec.children || []).forEach((cs, i) => { if (kids[i]) _sizing(kids[i], cs); });
}

function _place(node, spec) {
  const auto = spec.layoutMode && spec.layoutMode !== "NONE";
  const kids = node.children || [];
  (spec.children || []).forEach((cs, i) => {
    if (!kids[i]) return;
    if (!auto && typeof cs.x === "number") { kids[i].x = cs.x; kids[i].y = cs.y; }
    _place(kids[i], cs);
  });
}

const frame = _build(SPEC, true);
_page.appendChild(frame);
try { frame.resize(SPEC.width, SPEC.height); } catch (e) {}
_layout(frame, SPEC);
_sizing(frame, SPEC);
_place(frame, SPEC);

frame.x = compNode.x + compNode.width + _GAP;
frame.y = compNode.y;
frame.name = FRAME_NAME;

figma.currentPage.selection = _page.id === figma.currentPage.id ? [frame] : figma.currentPage.selection;
if (_page.id === figma.currentPage.id) figma.viewport.scrollAndZoomIntoView([frame]);

return {
  frameId: frame.id,
  pageId: _page.id,
  pageName: _page.name,
  nodesCreated: frame.findAll(() => true).length + 1,
  anchorsPresent: frame.findAll((n) => n.name.charAt(0) === "#").length,
  layoutFailures: _failures,
};

File: uspec/templates/frame.py
#!/usr/bin/env python3
"""Build one uSpec template as a detached frame next to a component — the on-demand route.

    python frame.py --template API --component <component-set-id> \
        --expect-file "<your Figma file>" --dry-run

This is the standalone driver for `frame.js`, used to test the substitution before it is wired
into the `create-*` skills. In a real render the skill injects `frame.js` itself, in place of
its "Import and Detach Template" step, and uses the returned `frameId` exactly as before.

Building WRITES to Figma, so `--yes-write` is mandatory and means the designer approved that
specific run. Afterwards the frame is read back and diffed against the captured description —
the same verification `replay.py` does.
"""

from __future__ import annotations

import argparse
import json
import sys
import time
from pathlib import Path

from replay import post, diff  # same transport and the same property-by-property comparison

def default_port() -> int:
    """This copy's port, from project.json — the same number the bridge binds."""
    try:
        sys.path.insert(0, str(Path(__file__).resolve().parents[2]))
        import project
        cfg = project.load()
        if cfg:
            return int(cfg["port"])
    except Exception:
        pass
    return 8787


HERE = Path(__file__).resolve().parent

READ_BACK = """
const _f = await figma.getNodeByIdAsync(FRAME_ID);
const _val = (v) => (v === figma.mixed ? "__MIXED__" : v);
const READ = ["visible","opacity","blendMode","layoutMode","layoutWrap","layoutAlign","layoutGrow",
  "layoutPositioning","primaryAxisAlignItems","counterAxisAlignItems","layoutSizingHorizontal",
  "layoutSizingVertical","itemSpacing","counterAxisSpacing","paddingTop","paddingRight",
  "paddingBottom","paddingLeft","topLeftRadius","topRightRadius","bottomLeftRadius",
  "bottomRightRadius","strokeAlign","strokeTopWeight","strokeRightWeight","strokeBottomWeight",
  "strokeLeftWeight","characters","fontSize","textAlignHorizontal","textAlignVertical",
  "textAutoResize","textCase","textDecoration","clipsContent"];
function readBack(n) {
  const o = { name: n.name, type: n.type, width: n.width, height: n.height };
  for (const p of READ) { const v = n[p]; if (v !== undefined && v !== null) o[p] = _val(v); }
  if (n.type === "TEXT" && n.fontName && n.fontName !== figma.mixed)
    o.fontName = { family: n.fontName.family, style: n.fontName.style };
  if (Array.isArray(n.fills)) o.fills = JSON.parse(JSON.stringify(n.fills));
  if (n.children) o.children = n.children.map(readBack);
  return o;
}
return readBack(_f);
"""


def main() -> None:
    p = argparse.ArgumentParser(description=__doc__, formatter_class=argparse.RawDescriptionHelpFormatter)
    p.add_argument("--template", required=True,
                   help="Screen reader | Color Annotation | Anatomy | API | Property | Structure | Motion")
    p.add_argument("--component", required=True, help="node id of the component being documented")
    p.add_argument("--expect-file", required=True)
    p.add_argument("--session")
    p.add_argument("--frame-name", help="defaults to '<component name> <template>'")
    p.add_argument("--gap", type=int, default=200)
    p.add_argument("--port", type=int, default=default_port(),
                   help="bridge port (default: this copy's project.json)")
    p.add_argument("--timeout", type=int, default=300)
    p.add_argument("--dry-run", action="store_true")
    p.add_argument("--yes-write", action="store_true",
                   help="required to build: confirms a go-ahead for this specific run")
    args = p.parse_args()

    src = json.loads((HERE / "templates.json").read_text(encoding="utf-8"))
    if args.template not in src["templates"]:
        sys.exit("unknown template %r. Have: %s" % (args.template, ", ".join(src["templates"])))
    spec = src["templates"][args.template]

    count = lambda o: 1 + sum(count(c) for c in o.get("children", []))
    anchors = lambda o: (1 if o["name"].startswith("#") else 0) + sum(anchors(c) for c in o.get("children", []))

    if args.dry_run or not args.yes_write:
        probe = (
            'const c = await figma.getNodeByIdAsync(COMPONENT_NODE_ID);\n'
            'if (figma.root.name !== EXPECT_FILE) throw new Error("wrong file: " + figma.root.name);\n'
            'if (!c) throw new Error("component not found: " + COMPONENT_NODE_ID);\n'
            'let p = c; while (p.parent && p.parent.type !== "DOCUMENT") p = p.parent;\n'
            'return { file: figma.root.name, component: c.name, page: p.name,\n'
            '  placeAt: { x: c.x + c.width + GAP, y: c.y } };\n'
        )
        v = post(args.port, probe,
                 {"COMPONENT_NODE_ID": args.component, "EXPECT_FILE": args.expect_file, "GAP": args.gap},
                 args.session, args.timeout)
        print("DRY RUN — nothing written to %s" % v["file"])
        print("  template : %s — %d nodes, %d #anchor layers" % (args.template, count(spec), anchors(spec)))
        print("  component: %s on page %r" % (v["component"], v["page"]))
        print("  frame at : x=%.0f y=%.0f" % (v["placeAt"]["x"], v["placeAt"]["y"]))
        if not args.dry_run:
            print("\n  Refusing to build without --yes-write.")
            sys.exit(1)
        return

    consts = {
        "SPEC": spec,
        "COMPONENT_NODE_ID": args.component,
        "FRAME_NAME": args.frame_name or ("%s %s" % (args.component, args.template)),
        "GAP": args.gap,
    }
    t0 = time.time()
    v = post(args.port, (HERE / "frame.js").read_text(encoding="utf-8"), consts, args.session, args.timeout)
    print("built %r on page %r in %.1f s" % (v["frameId"], v["pageName"], time.time() - t0))
    print("  %d nodes, %d #anchor layers (captured: %d / %d)"
          % (v["nodesCreated"], v["anchorsPresent"], count(spec), anchors(spec)))

    for f in v.get("layoutFailures") or []:
        print("  layout property rejected: %-24s %-22s = %-10r  %s"
              % (f["node"][:24], f["property"], f["value"], f["message"]))

    got = post(args.port, READ_BACK, {"FRAME_ID": v["frameId"]}, args.session, args.timeout)

    # The root's name is the frame name the render wants, not the template's; compare the rest.
    want = dict(spec)
    want["name"] = got.get("name")
    want["type"] = "FRAME"

    out: list = []
    diff(want, got, args.template, out)
    print("\nverifying — read back and diffed against templates.json")
    for path, prop, wv, gv in out[:20]:
        print("    %-52s %-22s want %r got %r" % (path[-52:], prop, wv, gv))
    if len(out) > 20:
        print("    ... %d more" % (len(out) - 20))
    print("\n%s — %d difference(s)" % ("VERIFIED" if not out else "MISMATCH", len(out)))
    print("Rollback: delete frame %s." % v["frameId"])
    if out:
        sys.exit(1)


if __name__ == "__main__":
    main()
